pyevmasm, tests: Add support for multiple forks.

cryptomental opened the pull request:
* separate instruction tables for pre-byzantium, byzantium and constantinopole.
* added new optional argument to CLI (--fork, -f) that accepts fork to be used.
* it is possible to select fork by name or by block number.
* introduced default fork selection by introducing DEFAULT_FORK constant
* added 3 new test cases, one per supported fork
* fixed test_ADD_1 test case

All functions and evmasm are fully backwards compatible. If no fork is selected, "byzantium" is chosen by default.

Example usage:

```
evmasm -t -f 0
evmasm -t -f 3700000
evmasm -t -f 4369999
evmasm -t -f pre-byzantium
evmasm -t
evmasm -t -f 4370000
evmasm -t -f 4370001
evmasm -t -f byzantium
evmasm -t -f 9999999  (block number to be updated after mainnet launch)
evmasm -t -f constantinople
```

Error handling:

```
$ evmasm -t -f a
Wrong fork name or block number. Please provide an integer or one of ['pre-byzantium', 'byzantium', 'constantinople'].
$ evmasm -t -f constantinoplee
Wrong fork name or block number. Please provide an integer or one of ['pre-byzantium', 'byzantium', 'constantinople'].
$ evmasm -t -f 1.2
Wrong fork name or block number. Please provide an integer or one of ['pre-byzantium', 'byzantium', 'constantinople'].
$ evmasm -t -f
usage: evmasm [-h] (-a | -d | -t) [-bi] [-bo] [-i [INPUT]] [-o [OUTPUT]]
              [-f FORK]
evmasm: error: argument -f/--fork: expected 1 argument
```

Refs: https://github.com/trailofbits/pyevmasm/issues/7

cryptomental commented:
Hi @dguido , @computerality before the review starts I would like to clarify that I introduced three instruction tables, one per each fork. It may look a bit redundant instead of patching pre-byzantium table, but might be error-prone in the future where there might be changes in gas costs, params etc. and I decided to implement this way. Feel free to modify it later if needed.

disconnect3d commented on pyevmasm/__main__.py:
```diff
@@ -18,9 +18,26 @@ def main():
     args = parser.parse_args(sys.argv[1:])
 
+    accepted_forks = ['pre-byzantium', 'byzantium', 'constantinople']
+    arg_fork = args.fork[0].lower()
```
```suggestion
    arg_fork = args.fork.lower()
```

cryptomental replied:
Hi @disconnect3d  that was my original implementation ;) When you set argparse required argument parameters to 1, it becomes an array.

disconnect3d commented on pyevmasm/__main__.py:
```diff
@@ -18,9 +18,26 @@ def main():
                         help='Input file, default=stdin')
     parser.add_argument('-o', '--output', nargs='?', default=sys.stdout, type=argparse.FileType('w'),
                         help='Output file, default=stdout')
+    parser.add_argument('-f', '--fork', nargs=1, default=DEFAULT_FORK, type=str,
```
```suggestion
    parser.add_argument('-f', '--fork', default=DEFAULT_FORK, type=str,
```

disconnect3d replied:
I assume there is no reason to specify a list of forks. When `nargs=1` is not passed, `args.fork` will return a string instead of a single item list with a string.

cryptomental replied:
Ok thanks. I will change it back to '?' (see my comment above).

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -24,7 +27,7 @@
         >>> disassemble_hex('0x606040526002610100')
         'PUSH1 0x60\\nBLOCKHASH\\nMSTORE\\nPUSH1 0x2\\nPUSH2 0x100'
-        
+
```
Hmm... I wonder if doctest would catch this string. If no, maybe it catches module docstrings and so moving it to a module docstring could let us launch doctest over it if we don't do it yet.

Idk, loud thinking, not relevant to the PR.

@computerality (marking u so u can decide if we want it and make an issue or skip it)

computerality replied:
made a separate issue to track this idea

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -122,8 +125,283 @@ def __setitem__(self, key, value):
+        0x19: ('NOT', 0, 1, 1, 3, 'Bitwise NOT operation.'),
+        0x1a: ('BYTE', 0, 2, 1, 3, 'Retrieve single byte from word.'),
+        0x20: ('SHA3', 0, 2, 1, 30, 'Compute Keccak-256 hash.'),
+        0x30: ('ADDRESS', 0, 0, 1, 2, 'Get address of currently executing account     .'),
```
```suggestion
        0x30: ('ADDRESS', 0, 0, 1, 2, 'Get address of currently executing account.'),
```

cryptomental replied:
Good catch. Will be corrected asap.

disconnect3d replied:
@cryptomental there is an `apply suggestion` button which should do the job! Cool new feature of github ;).

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -122,8 +125,283 @@ def __setitem__(self, key, value):
+        0x3a: ('GASPRICE', 0, 0, 1, 2, 'Get price of gas in current environment.'),
+        0x3b: ('EXTCODESIZE', 0, 1, 1, 20, "Get size of an account's code."),
+        0x3c: ('EXTCODECOPY', 0, 4, 0, 20, "Copy an account's code to memory."),
+        0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment'),
```
```suggestion
        0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment.'),
```

cryptomental replied:
Will be done.

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -122,8 +125,283 @@ def __setitem__(self, key, value):
+        0x3b: ('EXTCODESIZE', 0, 1, 1, 20, "Get size of an account's code."),
+        0x3c: ('EXTCODECOPY', 0, 4, 0, 20, "Copy an account's code to memory."),
+        0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment'),
+        0x3e: ('RETURNDATACOPY', 0, 3, 0, 3, 'Copy output data from the previous call to memory'),
```
```suggestion
        0x3e: ('RETURNDATACOPY', 0, 3, 0, 3, 'Copy output data from the previous call to memory.'),
```

cryptomental replied:
Will be done.

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -122,8 +125,283 @@ def __setitem__(self, key, value):
+        0x19: ('NOT', 0, 1, 1, 3, 'Bitwise NOT operation.'),
+        0x1a: ('BYTE', 0, 2, 1, 3, 'Retrieve single byte from word.'),
+        0x20: ('SHA3', 0, 2, 1, 30, 'Compute Keccak-256 hash.'),
+        0x30: ('ADDRESS', 0, 0, 1, 2, 'Get address of currently executing account     .'),
```
```suggestion
        0x30: ('ADDRESS', 0, 0, 1, 2, 'Get address of currently executing account.'),
```

cryptomental replied:
+1.

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -122,8 +125,283 @@ def __setitem__(self, key, value):
+        0x3a: ('GASPRICE', 0, 0, 1, 2, 'Get price of gas in current environment.'),
+        0x3b: ('EXTCODESIZE', 0, 1, 1, 20, "Get size of an account's code."),
+        0x3c: ('EXTCODECOPY', 0, 4, 0, 20, "Copy an account's code to memory."),
+        0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment'),
```
```suggestion
        0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment.'),
```

cryptomental replied:
+2

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -122,8 +125,283 @@ def __setitem__(self, key, value):
+        0x3b: ('EXTCODESIZE', 0, 1, 1, 20, "Get size of an account's code."),
+        0x3c: ('EXTCODECOPY', 0, 4, 0, 20, "Copy an account's code to memory."),
+        0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment'),
+        0x3e: ('RETURNDATACOPY', 0, 3, 0, 3, 'Copy output data from the previous call to memory'),
```
```suggestion
        0x3e: ('RETURNDATACOPY', 0, 3, 0, 3, 'Copy output data from the previous call to memory.'),
```

cryptomental replied:
+3 / I am wondering where did it came from...

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -147,6 +425,9 @@ def __setitem__(self, key, value):
+        0x1c: ('SHR', 0, 2, 1, 3, 'Logical right shift.'),
+        0x1d: ('SAR', 0, 2, 1, 3, 'Arithmetic right shift.'),
         0x20: ('SHA3', 0, 2, 1, 30, 'Compute Keccak-256 hash.'),
         0x30: ('ADDRESS', 0, 0, 1, 2, 'Get address of currently executing account     .'),
```
```suggestion
        0x30: ('ADDRESS', 0, 0, 1, 2, 'Get address of currently executing account.'),
```

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -163,6 +444,7 @@ def __setitem__(self, key, value):
         0x3c: ('EXTCODECOPY', 0, 4, 0, 20, "Copy an account's code to memory."),
         0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment'),
```
```suggestion
        0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment.'),
```

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -163,6 +444,7 @@ def __setitem__(self, key, value):
         0x3c: ('EXTCODECOPY', 0, 4, 0, 20, "Copy an account's code to memory."),
         0x3d: ('RETURNDATASIZE', 0, 0, 1, 2, 'Get size of output data from the previous call from the current environment'),
         0x3e: ('RETURNDATACOPY', 0, 3, 0, 3, 'Copy output data from the previous call to memory'),
```
```suggestion
        0x3e: ('RETURNDATACOPY', 0, 3, 0, 3, 'Copy output data from the previous call to memory.'),
```

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -255,11 +537,13 @@ def __setitem__(self, key, value):
         0xf2: ('CALLCODE', 0, 7, 1, 40, "Message-call into this account with alternative account's code."),
         0xf3: ('RETURN', 0, 2, 0, 0, 'Halt execution returning output data.'),
         0xf4: ('DELEGATECALL', 0, 6, 1, 40, "Message-call into this account with an alternative account's code, but persisting into this account with an alternative account's code."),
+        0xf5: ('CREATE2', 0, 4, 1, 32000, 'Skinny CREATE2.'),
```
Can we have a better description here?

cryptomental replied:
Yes, I will try to add a better one.

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -730,16 +1028,18 @@ def assemble(asmcode, pc=0):
             b"\x60\x60\x60\x40\x52\x60\x02\x61\x01\x00"
     """
-    return b''.join([x.bytes for x in assemble_all(asmcode, pc=pc)])
+    return b''.join([x.bytes for x in assemble_all(asmcode, pc=pc, fork=fork)])
```
```suggestion
    return b''.join(x.bytes for x in assemble_all(asmcode, pc=pc, fork=fork))
```

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -783,4 +1085,33 @@ def assemble_hex(asmcode, pc=0):
+    """
+    forks_by_block = {
+        0: "pre-byzantium",
+        4370000 : "byzantium",
```
```suggestion
        4370000: "byzantium",
```

disconnect3d commented on pyevmasm/evmasm.py:
```diff
@@ -783,4 +1085,33 @@ def assemble_hex(asmcode, pc=0):
+    forks_by_block = {
+        0: "pre-byzantium",
+        4370000 : "byzantium",
+        9999999 : "constantinople"  # to be replaced after Constantinopole launch
```
```suggestion
        9999999: "constantinople"  # to be replaced after Constantinopole launch
```

disconnect3d commented on tests/test_EVMAssembler.py:
```diff
@@ -44,6 +44,28 @@ def test_JUMPI(self):
+    def test_constantinople_fork(self):
+        insn = EVMAsm.disassemble_one(b'\x1d', fork='constantinople')
+        self.assertTrue(insn.mnemonic == 'SAR')
+        self.assertTrue(insn.is_arithmetic)
```
Can we also test all the others instructions that have been added?

cryptomental replied:
Yes. I thought about it as well. I will test all of them.

cryptomental replied:
All new instructions are tested now.

disconnect3d commented on tests/test_EVMAssembler.py:
```diff
@@ -44,6 +44,28 @@ def test_JUMPI(self):
+        insn = EVMAsm.disassemble_one(b'\x57', fork='pre-byzantium')
+        self.assertTrue(insn.mnemonic == 'JUMPI')
+        self.assertTrue(insn.is_branch)
+        insn = EVMAsm.disassemble_one(b'\xfd', fork='pre-byzantium')
```
Maybe we should iterate over all bytes that should be invalid and assert them.

cryptomental replied:
Done. All invalid are tested now.

disconnect3d commented on tests/test_EVMAssembler.py:
```diff
@@ -44,6 +44,28 @@ def test_JUMPI(self):
+        insn = EVMAsm.disassemble_one(b'\x1d', fork='byzantium')
+        self.assertTrue(insn.mnemonic == 'INVALID')
+        insn = EVMAsm.disassemble_one(b'\xf5', fork='byzantium')
+        self.assertTrue(insn.mnemonic == 'INVALID')
```
Can we add comments like in `# REVERT added in byzantium`?

cryptomental replied:
Ok.

cryptomental commented:
Hi @disconnect3d . Thank you for the detailed review. I will go back to the PR tomorrow morning.

cryptomental commented:
Hi @disconnect3d I addressed all findings, updated test cases as advised and amended the pull request. Tests still pass. ;) The pull request is ready for another round of the review.

cryptomental commented:
@disconnect3d I am on it asap!

cryptomental commented:
> @disconnect3d I am on it asap!

@disconnect3d I corrected instruction tables import in readme, docstrings and added new command line option description to readme. I did not touch anything else. Thanks for noticing this! I guess I will start my next PRs with checking README's ;)
